std/assembly/array.ts

```diff
@@ -493,6 +493,68 @@ export class Array<T> {
     return <string>unreachable();
   }
 
+  flat(): T {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    var length = this.length_;
+    var selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    var size = 0;
+    for (let i = 0; i < length; i++) {
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+      size += child == 0 ? 0 : load<i32>(child, offsetof<T>("length_"));
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    var byteLength = <usize>size << usize(alignof<valueof<T>>());
+    var dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value and initialize it
+    var result = __alloc(offsetof<T>(), idof<T>());
+    store<i32>(result, size, offsetof<T>("length_"));
+
+    // byteLength, dataStart, and buffer are all readonly
+    store<i32>(changetype<usize>(result), byteLength, offsetof<T>("byteLength"));
+    store<usize>(changetype<usize>(result), dataStart, offsetof<T>("dataStart"));
+    store<usize>(changetype<usize>(result), __retain(dataStart), offsetof<T>("buffer"));
+
+    // set the elements
+    var resultIndex = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+
+      // ignore null arrays
+      if (child == 0) continue;
+
+      // move to the next array index
+      resultIndex++;
+
+      // copy the underlying buffer data to the result buffer
+      let childDataLength = load<i32>(child, offsetof<T>("byteLength"));
+      memory.copy(
+        dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())),
+        load<usize>(child, offsetof<T>("dataStart")),
+        <usize>childDataLength,
+      );
+
+      // advance the result length
+      let childLength = childDataLength >> <i32>alignof<valueof<T>>();
+      resultIndex += childLength - 1;
+    }
+
+    // if the `valueof<T>` type is managed, we must call __retain() on each reference
+    if (isManaged<valueof<T>>()) {
+      for (let i = 0; i < size; i++) {
+        __retain(load<usize>(dataStart + (<usize>i << usize(alignof<valueof<T>>()))));
```

Comment:
```suggestion
        __retain(load<usize>(dataStart + (<usize>i << alignof<valueof<T>>())));
```

Reply:
This was to suppress a typescript error in the IDE.

Reply:
use `@ts-ignore`

Reply:
![image](https://user-images.githubusercontent.com/3761339/78272113-152aae80-74db-11ea-9212-c576ae3f35a9.png)

Even using ts-ignore, the problems don't go away.

@dcodeIO what kind of solution should we implement here?

Reply:
Looks like a TS bug, hmm. As if it'd interpret that as `i < <alignof` expecting an `<alignof>` type assertion. Strange.

Reply:
Seems to happen on `<<` only. Minimal repro:

```ts
function alignof<T>(): number {
  return 1;
}

function foo<T>(): number {
  return 1 << alignof<T[]>();
}
```

Might have something to do with how they rescan on potentially ambiguous `<` sequences that might be a type assertion or not.

Reply:
So it ts lang server issue?

Reply:
https://github.com/microsoft/TypeScript/issues/37760

Reply:
does this result in less optimal code, and should I find an alternative?

Reply:
Suggesting to leave it with broken syntax highlighting for now, if there are no other issues. Otherwise, like if there are also lint errors, perhaps add a TODO linking to the TS bug and use a workaround.